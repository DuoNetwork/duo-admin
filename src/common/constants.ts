export * from '../../../duo-contract-wrapper/src/constants';
export const TRADES = 'trades';
export const COMMIT = 'commit';
export const SUBSCRIBE = 'subscribe';
export const MINUTELY = 'minutely';
export const HOURLY = 'hourly';
export const NODE = 'node';
export const CLEAN_DB = 'cleanDB';
export const FETCH_PRICE = 'fetchPrice';
export const API_BITFINEX = 'bitfinex';
export const API_GEMINI = 'gemini';
export const API_KRAKEN = 'kraken';
export const API_GDAX = 'gdax';

export const AWS_DYNAMO_API_VERSION = '2012-10-08';
// db setting
export const DB_SQL_SCHEMA_PRICEFEED = 'priceFeedDB';
export const DB_SQL_TRADE = 'trades';
export const DB_SQL_HISTORY = 'historical_price';

export const DB_DUO = 'duo';
export const DB_TRADES = 'trades';
export const DB_LIVE = 'live';
export const DB_DEV = 'dev';
export const DB_PRICES = 'prices';
export const DB_AWS_TRADES_LIVE = 'duo.trades.live';
export const DB_AWS_TRADES_DEV = 'duo.trades.dev';
export const DB_AWS_HOURLY_LIVE = 'duo.prices.60.live';
export const DB_AWS_HOURLY_DEV = 'duo.prices.60.dev';
export const DB_AWS_MINUTELY_LIVE = 'duo.prices.1.live';
export const DB_AWS_MINUTELY_DEV = 'duo.prices.1.dev';
export const DB_AWS_EVENTS_LIVE = 'duo.events.live';
export const DB_AWS_EVENTS_DEV = 'duo.events.dev';
export const DB_AWS_STATUS_LIVE = 'duo.status.live';
export const DB_AWS_STATUS_DEV = 'duo.status.dev';
export const DB_AWS_UI_EVENTS_LIVE = 'duo.uiEvents.live';
export const DB_AWS_UI_EVENTS_DEV = 'duo.uiEvents.dev';
export const DB_TX_QTE = 'quote';
export const DB_TX_BASE = 'base';
export const DB_TX_QUOTE_BASE_ID = 'quoteBaseId';
export const DB_TX_SRC = 'source';
export const DB_TX_ID = 'id';
export const DB_TX_PRICE = 'price';
export const DB_TX_AMOUNT = 'amount';
export const DB_TX_TS = 'timestamp';
export const DB_TX_SYSTIME = 'systime';
export const DB_TX_PAIR = 'pair';
export const DB_UPDATED_AT = 'updatedAt';
export const DB_TX_SRC_DHM = 'sourceDateHourMinute';
export const DB_HISTORY_PRICE = 'price';
export const DB_HISTORY_TIMESTAMP = 'timestamp';
export const DB_QUOTE_BASE_TS = 'quoteBaseTimestamp';
export const DB_HISTORY_VOLUME = 'volume';
export const DB_SRC_DHM = 'sourceDateHourMinute';
export const DB_SRC_DH = 'sourceDateHour';
export const DB_SRC_DATE = 'sourceDate';
export const DB_SRC_YM = 'sourceYearMonth';
export const DB_MN_MINUTE = 'minute';
export const DB_OHLC_OPEN = 'open';
export const DB_OHLC_HIGH = 'high';
export const DB_OHLC_LOW = 'low';
export const DB_OHLC_CLOSE = 'close';
export const DB_OHLC_VOLUME = 'volume';
export const DB_OHLC_TS = 'timestamp';
export const DB_ST_PROCESS = 'process';
export const DB_ST_TS = 'timestamp';
export const DB_ST_BLOCK = 'block';
export const DB_ST_SYSTIME = 'systime';
export const DB_EV_KEY = 'eventKey';
export const DB_EV_TIMESTAMP_ID = 'timestampId';
export const DB_EV_SYSTIME = 'systime';
export const DB_EV_BLOCK_HASH = 'blockHash';
export const DB_EV_BLOCK_NO = 'blockNumber';
export const DB_EV_TX_HASH = 'transactionHash';
export const DB_EV_LOG_STATUS = 'logStatus';
export const DB_EV_PX = 'priceInWei';
export const DB_EV_TS = 'timeInSecond';
export const DB_EV_NAV_A = 'navAInWei';
export const DB_EV_NAV_B = 'navBInWei';
export const DB_EV_TOKEN_A = 'tokenAInWei';
export const DB_EV_TOKEN_B = 'tokenBInWei';
export const DB_EV_ETH = 'ethAmtInWei';
export const DB_EV_ETH_FEE = 'ethFeeInWei';
export const DB_EV_DUO_FEE = 'duoFeeInWei';
export const DB_EV_TOTAL_SUPPLY_A = 'totalSupplyA';
export const DB_EV_TOTAL_SUPPLY_B = 'totalSupplyB';
export const DB_EV_UI_ETH = 'eth';
export const DB_EV_UI_TOKEN_A = 'tokenA';
export const DB_EV_UI_TOKEN_B = 'tokenB';
export const DB_EV_UI_ETH_FEE = 'ethFee';
export const DB_EV_UI_DUO_FEE = 'duoFee';
export const DB_STATUS_EVENT_PUBLIC_OTHERS = 'EVENT_AWS_PUBLIC_OTHERS';

export const STATUS_INTERVAL = 30;
export const TRADES_STATUS_LAST_UPDATE_INTERVAL_WS = 10;

export const AWS_DYNAMO_ROLE_TRADE = 'trade';
export const AWS_DYNAMO_ROLE_HOURLY = 'hourly';
export const AWS_DYNAMO_ROLE_MINUTELY = 'minutely';
export const AWS_DYNAMO_ROLE_EVENT = 'event';
export const AWS_DYNAMO_ROLE_STATUS = 'status';

export const API_FETCH_TRADE_INTERVAL = 1;
export const LOG_INFO = 'INFO';
export const LOG_ERROR = 'ERROR';
export const LOG_DEBUG = 'DEBUG';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const DB_PRICES_PRIMARY_KEY_RESOLUTION: {
	[period: number]: 'minute' | 'hour' | 'day' | 'month';
} = {
	0: 'minute',
	1: 'hour',
	10: 'hour',
	60: 'day',
	360: 'month',
	1440: 'month'
};

export const API_GMN_BASE_URL = 'https://api.gemini.com';
export const API_GMN_WS_LINK = 'wss://api.gemini.com';
export const API_GMN_VERSION = '/v1';
export const API_GMN_ORDER_BOOK = '/book';
export const API_GMN_TRADE = '/trades';
export const API_BFX_BASE_URL = 'https://api.bitfinex.com';
export const API_BFX_VERSION = '/v1';
export const API_BFX_ORDER_BOOK = '/book/';
export const API_BFX_TRADE = '/trades/';
export const API_KRK_BASE_URL = 'https://api.kraken.com';
export const API_KRK_VERSION = '/0/public';
export const API_KRK_ORDER_BOOK = '/Depth';
export const API_KRK_TRADE = '/Trades';
export const API_GDAX_BASE_URL = 'https://api.gdax.com:443/products';
export const API_GDAX_TRADE = '/trades';

export const API_LIST = [API_KRAKEN, API_GEMINI, API_GDAX];
export const EXCHANGE_WEIGHTAGE_TH: { [index: number]: number[] } = {
	4: [0.35, 0.3, 0.25, 0.2],
	3: [0.55, 0.45, 0.35],
	2: [0.65, 0.5]
};
