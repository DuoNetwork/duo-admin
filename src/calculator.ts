import sqlUtil from './sqlUtil';
import * as CST from './constants';
import { Price, Trade } from './types';

export class Calculateor {
	getTradesForInterval(trades: Trade[], currentTimestamp: number, interval: number): Trade[] {
		const oneInterval = 5 * 60 * 1000;
		const upperTime: number = currentTimestamp - interval * oneInterval;
		const lowerTime: number = upperTime - oneInterval;

		return trades.filter(
			trade =>
				Number(trade.timestamp) <= upperTime &&
				Number(trade.timestamp) > lowerTime
		);
	}

	getVolumeMedianPrice(trades: Trade[], timestamp: number): Price {
		trades.sort((a, b) => Number(a.price) - Number(b.price));
		const cumVols: number[] = [];
		let cumVol = 0;
		trades.forEach((trade, index) => {
			cumVol += Number(trade.amount);
			cumVols[index] = cumVol;
		});
		const halfTotalVol: number = cumVol / 2;
		const medianIndex = cumVols.findIndex(v => v >= halfTotalVol);
		return { price: Number(trades[medianIndex].price), volume: cumVol, timestamp: timestamp };
	}

	getExchangePriceFix(trades: Trade[], currentTimestamp: number): Price {
		for (let i = 0; i < 12; i++) {
			const subTrades = this.getTradesForInterval(trades, currentTimestamp, i);
			if (subTrades.length > 0) return this.getVolumeMedianPrice(subTrades, currentTimestamp);
		}

		return { price: 0, volume: 0, timestamp: currentTimestamp };
	}

	getWeights(rawVolume: number[]): number[] {
		const totalVol = rawVolume.reduce((a, b) => a + b);
		const weightArray = rawVolume.map(v => v / totalVol);
		return weightArray;
	}

	validateWeights(weights: number[]): boolean {
		const numOfValidExchanges = weights.length;
		for (let i = 0; i < weights.length; i++)
			if (weights[i] > CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges][i]) return false;
		return true;
	}

	modifyWeights(weights: number[]): number[] {
		const numOfValidExchanges = weights.length;
		let isValid: boolean = this.validateWeights(weights);
		while (!isValid) {
			let sumOfCapped: number = 0;
			let sumOfUnCapped: number = 0;
			const weightCaps = CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges];
			for (let i = 0; i < weights.length; i++) {
				if (weights[i] >= weightCaps[i]) {
					weights[i] = weightCaps[i];
					sumOfCapped += weightCaps[i];
				} else {
					sumOfUnCapped += weights[i];
				}
			}

			for (let i = 0; i < weights.length; i++) {
				if (weights[i] < weightCaps[i]) {
					weights[i] = weights[i] / sumOfUnCapped * (1 - sumOfCapped);
				}
			}
			// console.log(weights);
			isValid = this.validateWeights(weights);
		}
		return weights;
	}

	consolidatePriceFix(exchangePriceVolume: Array<Price>): number {
		const filterredExchanges = exchangePriceVolume.filter(item => item.volume > 0);

		// sort based on volume from large to small
		filterredExchanges.sort((a, b) => b.volume - a.volume);
		const volumeList = filterredExchanges.map(item => item.volume);

		if (filterredExchanges.length === 0)
			// use previous priceFix
			return 0;
		else if (filterredExchanges.length === 1)
			// let finalArray: number[] = [];
			return filterredExchanges[0].price;
		else {
			console.log('there are ' + filterredExchanges.length + ' valid exchanges');
			const totalVol = volumeList.reduce((a, b) => a + b);
			const weights = volumeList.map(v => v / totalVol);
			const finalWeights: number[] = this.modifyWeights(weights);
			return finalWeights.reduce(
				(accum, w, index) => accum + exchangePriceVolume[index].price * w
			);
		}
	}

	async getPriceFix(): Promise<Price> {
		const currentTimestamp: number = Math.floor(Date.now());
		const trades = await sqlUtil.readSourceData(currentTimestamp);
		const EXCHANGES_TRADES: { [key: string]: Trade[] } = {
			[CST.EXCHANGE_BITFINEX]: [],
			[CST.EXCHANGE_GEMINI]: [],
			[CST.EXCHANGE_GDAX]: [],
			[CST.EXCHANGE_KRAKEN]: []
		};

		trades.forEach(item => {
			if (item.source === CST.EXCHANGE_BITFINEX)
				EXCHANGES_TRADES[CST.EXCHANGE_BITFINEX].push(item);
			else if (item.source === CST.EXCHANGE_GEMINI)
				EXCHANGES_TRADES[CST.EXCHANGE_GEMINI].push(item);
			else if (item.source === CST.EXCHANGE_GDAX)
				EXCHANGES_TRADES[CST.EXCHANGE_GDAX].push(item);
			else if (item.source === CST.EXCHANGE_KRAKEN)
				EXCHANGES_TRADES[CST.EXCHANGE_KRAKEN].push(item);
		});

		const exchangePriceVolume = CST.EXCHANGES.map(src =>
			this.getExchangePriceFix(EXCHANGES_TRADES[src], currentTimestamp)
		);

		const priceFix: number = this.consolidatePriceFix(exchangePriceVolume);

		if (priceFix === 0) {
			console.log('no priceFix found, use the last ETH price');
			const lastPriceObj = await sqlUtil.readLastPrice();
			console.log(
				'the priceFix is: ' + lastPriceObj.price + ' at timestamp ' + lastPriceObj.timestamp
			);
			return lastPriceObj;
		} else {
			console.log('the priceFix is: ' + priceFix + ' at timestamp ' + currentTimestamp);
			console.log(exchangePriceVolume);
			console.log(exchangePriceVolume.reduce((sum, p) => sum + p.volume, 0));
			const priceObj = {
				price: priceFix,
				volume: exchangePriceVolume.reduce((sum, p) => sum + p.volume, 0),
				timestamp: currentTimestamp
			};
			// save price into DB
			await sqlUtil.insertPrice(priceObj);

			return priceObj;
		}
	}
}
const calculator = new Calculateor();
export default calculator;
