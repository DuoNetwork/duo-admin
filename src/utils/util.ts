import moment from 'moment';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
// import { Options } from 'aws-cli-js';

class Util {
	public logLevel: string = CST.LOG_INFO;

	private log(text: any, level?: string): void {
		if (level && CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	public composeQuery(paras: { [key: string]: any }): string {
		let query: string = '?';
		for (const key in paras) query += key + '=' + paras[key] + '&';
		query = query.slice(0, -1);
		return query;
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			forceREST: argv.includes('rest'),
			live: process.argv.includes('live'),
			dbLive: process.argv.includes('dbLive'),
			server: process.argv.includes('server'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			pair: '',
			contractType: 'Beethoven',
			tenor: 'Perpetual',
			assets: [''],
			sources: [''],
			exSources: [''],
			gasPrice: 5e9,
			gasLimit: 200000,
			source: '',
			event: '',
			provider: '',
			period: 1,
			base: 'USD',
			quote: 'ETH'
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'exSources':
					option.exSources = args[1].split(',');
					break;
				case 'sources':
					option.sources = args[1].split(',');
					break;
				case 'pair':
					option.pair = args[1].replace('_', '|') || option.pair;
					break;
				case 'assets':
					option.assets = args[1].split(',');
					break;
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'period':
					option.period = Number(args[1]);
					break;
				case 'base':
					option.base = args[1] || option.base;
					break;
				case 'quote':
					option.quote = args[1] || option.quote;
					break;
				case 'contractType':
					option.contractType = args[1] || option.contractType;
					break;
				case 'tenor':
					option.tenor = args[1] || option.tenor;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public getPeriodStartTimestamp(timestamp: number, period: number = 1) {
		// const now = this.getUTCNowTimestamp();
		return Math.floor(timestamp / 60000 / period - 1) * 60000 * period;
	}

	public timestampToString(ts: number) {
		//just for debugging purpose
		return moment.utc(ts).format('YYYY-MM-DDTHH:mm:ss');
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		console.log('dynamo', option.dynamo);
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case CST.TRADES:
				type = 'TRADE';
				source = option.source ? '_' + option.source.toUpperCase() : '';
				break;
			case CST.TRIGGER:
			case CST.FETCH_EVENTS:
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case CST.COMMIT:
				type = 'FEED';
				break;
			case CST.CLEAN_DB:
				type = 'CLEANDB';
				break;
			case CST.DB_PRICES:
				type = 'PRICE';
				if (option.period === 1) source = '_MINUTELY';
				else if (option.period === 60) source = '_HOURLY';
				else source = '';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case CST.FETCH_PRICE:
				type = 'FETCH';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public async sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}
}

const util = new Util();
export default util;
