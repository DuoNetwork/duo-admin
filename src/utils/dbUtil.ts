import Web3Wrapper from '../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from '../common/constants';
import { IOption, IPriceFix, ITrade } from '../common/types';
import dynamoUtil from './dynamoUtil';
import keyUtil from './keyUtil';
import sqlUtil from './sqlUtil';
import util from './util';

class DbUtil {
	public dynamo: boolean = false;

	public async init(tool: string, option: IOption, web3Wrapper: Web3Wrapper) {
		this.dynamo = option.dynamo;
		const process = util.getStatusProcess(tool, option);
		util.logInfo('process: ' + process);

		let config = {};
		try {
			config = require('../keys/aws/' + (option.live ? 'live' : 'dev') + '/admin.json');
		} catch (error) {
			util.logError(error);
		}
		await dynamoUtil.init(
			config,
			option.live,
			process,
			(value: string | number) => web3Wrapper.fromWei(value),
			async txHash => {
				const txReceipt = await web3Wrapper.getTransactionReceipt(txHash);
				if (!txReceipt) return null;
				return {
					status: txReceipt.status
				};
			}
		);
		if ([CST.TRADES, CST.COMMIT, CST.CLEAN_DB].includes(tool) && !option.dynamo) {
			const sqlAuth = await keyUtil.getSqlAuth(option);
			return sqlUtil.init(sqlAuth.host, sqlAuth.user, sqlAuth.password);
		}
	}

	public insertTradeData(trade: ITrade, insertStatus: boolean) {
		return this.dynamo
			? dynamoUtil.insertTradeData(trade, insertStatus)
			: sqlUtil.insertTradeData(trade, insertStatus);
	}

	public insertPrice(price: IPriceFix) {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.insertPrice(price);
	}

	public readLastPrice(quote: string, base: string): Promise<IPriceFix> {
		return this.dynamo ? Promise.reject('invalid') : sqlUtil.readLastPrice(quote, base);
	}

	public readSourceData(
		currentTimestamp: number,
		quote: string,
		base: string
	): Promise<ITrade[]> {
		return this.dynamo
			? Promise.reject('invalid')
			: sqlUtil.readSourceData(currentTimestamp, base, quote);
	}

	public cleanDB(): Promise<void> {
		return this.dynamo ? Promise.reject() : sqlUtil.cleanDB();
	}

	public insertHeartbeat(data: object = {}): Promise<void> {
		return dynamoUtil.insertHeartbeat(data);
	}
}

const dbUtil = new DbUtil();
export default dbUtil;
