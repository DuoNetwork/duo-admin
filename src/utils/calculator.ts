import {Constants, IPriceFix, ITrade} from '@finbook/duo-market-data'
import * as CST from '../common/constants';
import dbUtil from './dbUtil';
import util from './util';

class Calculateor {
	public getTradesForInterval(
		trades: ITrade[],
		currentTimestamp: number,
		interval: number
	): ITrade[] {
		const oneInterval = 5 * 60 * 1000;
		const upperTime: number = currentTimestamp - interval * oneInterval;
		const lowerTime: number = upperTime - oneInterval;

		return trades.filter(
			trade => Number(trade.timestamp) <= upperTime && Number(trade.timestamp) > lowerTime
		);
	}

	public getVolumeMedianPrice(trades: ITrade[], timestamp: number): IPriceFix {
		trades.sort((a, b) => Number(a.price) - Number(b.price));
		const cumVols: number[] = [];
		let cumVol = 0;
		trades.forEach((trade, index) => {
			cumVol += Number(trade.amount);
			cumVols[index] = cumVol;
		});
		const halfTotalVol: number = cumVol / 2;
		const medianIndex = cumVols.findIndex(v => v >= halfTotalVol);
		return {
			price: Number(trades[medianIndex].price),
			volume: cumVol,
			timestamp,
			source: '',
			base: trades[0].base,
			quote: trades[0].quote
		};
	}

	public getExchangePriceFix(trades: ITrade[], currentTimestamp: number): IPriceFix {
		for (let i = 0; i < 12; i++) {
			const subTrades = this.getTradesForInterval(trades, currentTimestamp, i);
			if (subTrades.length > 0) return this.getVolumeMedianPrice(subTrades, currentTimestamp);
		}

		return {
			price: 0,
			volume: 0,
			timestamp: currentTimestamp,
			source: '',
			base: '',
			quote: ''
		};
	}

	public getWeights(rawVolume: number[]): number[] {
		const totalVol = rawVolume.reduce((a, b) => a + b, 0);
		const weightArray = rawVolume.map(v => v / totalVol);
		return weightArray;
	}

	public validateWeights(weights: number[]): boolean {
		const numOfValidExchanges = weights.length;
		for (let i = 0; i < weights.length; i++)
			if (weights[i] > CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges][i]) return false;

		return true;
	}

	public modifyWeights(weights: number[]): number[] {
		const numOfValidExchanges = weights.length;
		let isValid: boolean = this.validateWeights(weights);
		while (!isValid) {
			let sumOfCapped: number = 0;
			let sumOfUnCapped: number = 0;
			const weightCaps = CST.EXCHANGE_WEIGHTAGE_TH[numOfValidExchanges];
			for (let i = 0; i < weights.length; i++)
				if (weights[i] >= weightCaps[i]) {
					weights[i] = weightCaps[i];
					sumOfCapped += weightCaps[i];
				} else sumOfUnCapped += weights[i];

			for (let i = 0; i < weights.length; i++)
				if (weights[i] < weightCaps[i])
					weights[i] = (weights[i] / sumOfUnCapped) * (1 - sumOfCapped);

			isValid = this.validateWeights(weights);
		}
		return weights;
	}

	public consolidatePriceFix(exchangePriceVolume: IPriceFix[]): number {
		const filterredExchanges = exchangePriceVolume.filter(item => item.volume > 0);

		// sort based on volume from large to small
		filterredExchanges.sort((a, b) => b.volume - a.volume);
		const volumeList = filterredExchanges.map(item => item.volume);

		if (filterredExchanges.length === 0)
			// use previous priceFix
			return 0;
		else if (filterredExchanges.length === 1)
			// let finalArray: number[] = [];
			return filterredExchanges[0].price;
		else {
			util.logInfo('there are ' + filterredExchanges.length + ' valid exchanges');
			const totalVol = volumeList.reduce((a, b) => a + b, 0);
			const weights = volumeList.map(v => v / totalVol);
			const finalWeights: number[] = this.modifyWeights(weights);
			return finalWeights.reduce(
				(accum, w, index) => accum + filterredExchanges[index].price * w,
				0
			);
		}
	}

	public async getPriceFix(quote: string, base: string): Promise<IPriceFix> {
		const currentTimestamp: number = util.getUTCNowTimestamp();
		const trades = await dbUtil.readSourceData(currentTimestamp, base, quote);
		const EXCHANGES_TRADES: { [key: string]: ITrade[] } = {
			[Constants.API_BITFINEX]: [],
			[Constants.API_GEMINI]: [],
			[Constants.API_GDAX]: [],
			[Constants.API_KRAKEN]: [],
			[Constants.API_BITSTAMP]: []
		};

		trades.forEach(item => {
			if (item.source === Constants.API_BITFINEX)
				EXCHANGES_TRADES[Constants.API_BITFINEX].push(item);
			else if (item.source === Constants.API_GEMINI)
				EXCHANGES_TRADES[Constants.API_GEMINI].push(item);
			else if (item.source === Constants.API_GDAX)
				EXCHANGES_TRADES[Constants.API_GDAX].push(item);
			else if (item.source === Constants.API_KRAKEN)
				EXCHANGES_TRADES[Constants.API_KRAKEN].push(item);
			else if (item.source === Constants.API_BITSTAMP)
				EXCHANGES_TRADES[Constants.API_BITSTAMP].push(item);
		});

		const exchangePriceVolume = CST.API_LIST.map(src =>
			this.getExchangePriceFix(EXCHANGES_TRADES[src], currentTimestamp)
		);

		const priceFix: number = this.consolidatePriceFix(exchangePriceVolume);
		util.logInfo('priceFix calculated is ' + priceFix);

		if (priceFix === 0) {
			util.logInfo('no priceFix found, use the last ETH price');
			const lastPriceObj = await dbUtil.readLastPrice(base, quote);
			util.logInfo(
				'the priceFix is: ' + lastPriceObj.price + ' at timestamp ' + lastPriceObj.timestamp
			);
			return lastPriceObj;
		} else {
			const priceObj = {
				price: priceFix,
				volume: exchangePriceVolume.reduce((sum, p) => sum + p.volume, 0),
				timestamp: currentTimestamp,
				source: '',
				base: base,
				quote: quote
			};
			util.logInfo(
				'valid exchange priceFix found: ' +
					priceObj.price +
					' at ' +
					priceObj.timestamp +
					' for volume of ' +
					priceObj.volume
			);
			// save price into DB
			await dbUtil.insertPrice(priceObj);

			return priceObj;
		}
	}
}
const calculator = new Calculateor();
export default calculator;
