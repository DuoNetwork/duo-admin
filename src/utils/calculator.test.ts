// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import * as CST from '../common/constants';
import { ITrade } from '../common/types';
import trades from '../samples/ETHUSDtrades.json';
import trades2 from '../samples/ETHUSDtrades2.json';
import calculator from './calculator';
import dbUtil from './dbUtil';
import util from './util';

test('getVolumeMedianPrice', () => {
	CST.API_LIST.forEach(exchange => {
		const exchangeTrades: ITrade[] = trades.filter(item => item.source === exchange);
		expect(calculator.getVolumeMedianPrice(exchangeTrades, 1234567890)).toMatchSnapshot();
	});
});

test('modifyWeights', () => {
	let inputWeightage: number[] = [0.5, 0.4, 0.05, 0.05];
	expect(calculator.modifyWeights(inputWeightage)).toMatchSnapshot();

	inputWeightage = [0.7, 0.2, 0.1];
	expect(calculator.modifyWeights(inputWeightage)).toMatchSnapshot();

	inputWeightage = [0.8, 0.2];
	expect(calculator.modifyWeights(inputWeightage)).toMatchSnapshot();
});

test('getExchangePriceFix', () => {
	CST.API_LIST.forEach(exchange => {
		const exchangeTrades: ITrade[] = trades.filter(item => item.source === exchange);
		const timestamp = exchangeTrades.reduce(
			(min, p) => (Number(p.timestamp) < min ? Number(p.timestamp) : min),
			Number(exchangeTrades[0].timestamp)
		);
		expect(calculator.getExchangePriceFix(exchangeTrades, timestamp)).toMatchSnapshot();
	});
});

test('getPriceFix case 1', async () => {
	dbUtil.readSourceData = jest.fn(() => Promise.resolve(trades));
	dbUtil.insertPrice = jest.fn(() => Promise.resolve());
	util.getUTCNowTimestamp = jest.fn(() => 1524547909941);
	await calculator.getPriceFix('ETH', 'USD');
	expect((dbUtil.insertPrice as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

test('getPriceFix case 2', async () => {
	dbUtil.readSourceData = jest.fn(() => Promise.resolve(trades2));
	dbUtil.insertPrice = jest.fn(() => Promise.resolve());
	util.getUTCNowTimestamp = jest.fn(() => 1524547909941);
	await calculator.getPriceFix('ETH', 'USD');
	expect((dbUtil.insertPrice as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

test('getPriceFix, no priceFix', async () => {
	dbUtil.readSourceData = jest.fn(() => Promise.resolve([]));
	dbUtil.insertPrice = jest.fn(() => Promise.resolve());
	util.getUTCNowTimestamp = jest.fn(() => 1524547909941);
	dbUtil.readLastPrice = jest.fn(() =>
		Promise.resolve({
			price: 100,
			volume: 1000
		})
	);
	await calculator.getPriceFix('ETH', 'USD');
	expect((dbUtil.readSourceData as jest.Mock<Promise<void>>).mock.calls).toMatchSnapshot();
});

test('getWeights', () => {
	expect(calculator.getWeights([1, 2, 3, 4])).toMatchSnapshot();
});

test('consolidatePriceFix', () => {
	expect(
		calculator.consolidatePriceFix([
			{
				source: 'source',
				base: 'base',
				quote: 'quote',
				timestamp: 1234567890000,
				price: 100,
				volume: 10000
			}
		])
	).toMatchSnapshot();
});
