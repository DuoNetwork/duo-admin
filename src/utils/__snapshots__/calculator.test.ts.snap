// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExchangePriceFix 1`] = `
Object {
  "base": "USD",
  "price": 634.7,
  "quote": "ETH",
  "source": "",
  "timestamp": 1524453966393,
  "volume": 0.4,
}
`;

exports[`getExchangePriceFix 2`] = `
Object {
  "base": "USD",
  "price": 637,
  "quote": "ETH",
  "source": "",
  "timestamp": 1524465953111,
  "volume": 6.002117,
}
`;

exports[`getExchangePriceFix 3`] = `
Object {
  "base": "USD",
  "price": 667,
  "quote": "ETH",
  "source": "",
  "timestamp": 1524547898561,
  "volume": 0.024,
}
`;

exports[`getExchangePriceFix 4`] = `
Object {
  "base": "USD",
  "price": 633,
  "quote": "ETH",
  "source": "",
  "timestamp": 1524463065507,
  "volume": 8.67,
}
`;

exports[`getPriceFix case 1 1`] = `
Object {
  "base": "USD",
  "price": 666.98,
  "quote": "ETH",
  "source": "",
  "timestamp": 1524547909941,
  "volume": 0.16599999999999998,
}
`;

exports[`getPriceFix case 2 1`] = `
Object {
  "base": "USD",
  "price": 638.5705,
  "quote": "ETH",
  "source": "",
  "timestamp": 1524547909941,
  "volume": 35.73263379,
}
`;

exports[`getVolumeMedianPrice 1`] = `
Object {
  "base": "USD",
  "price": 634.69,
  "quote": "ETH",
  "source": "",
  "timestamp": 1234567890,
  "volume": 3.15308079,
}
`;

exports[`getVolumeMedianPrice 2`] = `
Object {
  "base": "USD",
  "price": 638,
  "quote": "ETH",
  "source": "",
  "timestamp": 1234567890,
  "volume": 22.681753,
}
`;

exports[`getVolumeMedianPrice 3`] = `
Object {
  "base": "USD",
  "price": 666.98,
  "quote": "ETH",
  "source": "",
  "timestamp": 1234567890,
  "volume": 0.16599999999999998,
}
`;

exports[`getVolumeMedianPrice 4`] = `
Object {
  "base": "USD",
  "price": 633,
  "quote": "ETH",
  "source": "",
  "timestamp": 1234567890,
  "volume": 12.835600000000001,
}
`;

exports[`modifyWeights 1`] = `
Array [
  0.35,
  0.3,
  0.17500000000000004,
  0.17500000000000004,
]
`;

exports[`modifyWeights 2`] = `
Array [
  0.55,
  0.29999999999999993,
  0.14999999999999997,
]
`;

exports[`modifyWeights 3`] = `
Array [
  0.65,
  0.35,
]
`;
